File: AspNetMvcUnleashed/0672329980_source/Chapter05/VB/EFMvcApplication/EFMvcApplication.Tests/Controllers/HomeControllerTestMock.vb
Imports Microsoft.VisualStudio.TestTools.UnitTesting
Imports Moq
Imports System.Web.Mvc
Imports GenericRepository

<TestClass()> _
Public Class HomeControllerTestMock

    Private _mockRepository As Mock(Of IGenericRepository)

    <TestInitialize()> _
    Sub Initialize()
        _mockRepository = New Mock(Of IGenericRepository)()
    End Sub


    <TestMethod()> _
    Sub NameIsRequired()
        ' Arrange
        Dim controller As New HomeController(_mockRepository.Object)
        Dim productToCreate As New Product()
        productToCreate.Name = String.Empty

        ' Act
        Dim result As ViewResult = controller.Create(productToCreate)

        ' Assert
        Dim modelStateError = result.ViewData.ModelState("Name").Errors(0).ErrorMessage
        Assert.AreEqual("Product name is required.", modelStateError)
    End Sub

End Class

File: AspNetMvcUnleashed/0672329980_source/Chapter05/CS/EFMvcApplication/EFMvcApplication.Tests/Controllers/HomeControllerTestMock.cs
using System;
using System.Web.Mvc;
using EFMvcApplication.Controllers;
using EFMvcApplication.Models;
using GenericRepository;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Moq;

namespace EFMvcApplication.Tests.Controllers
{
    [TestClass]
    public class HomeControllerTestMock
    {
        private Mock<IGenericRepository> _mockRepository;

        [TestInitialize]
        public void Initialize()
        {
            _mockRepository = new Mock<IGenericRepository>();
        }

        [TestMethod]
        public void NameIsRequired()
        {
            // Arrange
            var controller = new HomeController(_mockRepository.Object);
            var productToCreate = new Product();
            productToCreate.Name = String.Empty;

            // Act
            var result = (ViewResult)controller.Create(productToCreate);

            // Assert
            var modelStateError = result.ViewData.ModelState["Name"].Errors[0].ErrorMessage;
            Assert.AreEqual("Product name is required.", modelStateError);
        }
    }
}
